File: src/dotnet/Ansight.Core/Pairing/PairingFileTransferWireProtocol.cs
using System.Buffers.Binary;
using System.Text;

namespace Ansight.Pairing;

internal enum PairingFileTransferFrameType : byte
{
    Chunk = 1,
    Complete = 2,
    Error = 3
}

internal static class PairingFileTransferWireProtocol
{
    internal const string ProtocolName = "ansight.file-transfer.v1";
    internal const int HeaderSize = 56;
    private const byte Version = 1;

    internal static byte[] CreateFrame(
        Guid transferId,
        PairingFileTransferFrameType frameType,
        int sequence,
        long offsetBytes,
        ReadOnlySpan<byte> payload)
    {
        var frame = new byte[HeaderSize + payload.Length];
        WriteHeader(frame, transferId, frameType, sequence, offsetBytes, payload.Length);
        payload.CopyTo(frame.AsSpan(HeaderSize));
        return frame;
    }

    internal static void WriteHeader(
        Span<byte> header,
        Guid transferId,
        PairingFileTransferFrameType frameType,
        int sequence,
        long offsetBytes,
        int payloadByteCount)
    {
        if (header.Length < HeaderSize)
        {
            throw new ArgumentException("The file transfer header buffer was too small.", nameof(header));
        }

        header[0] = (byte)'A';
        header[1] = (byte)'S';
        header[2] = (byte)'F';
        header[3] = (byte)'T';
        header[4] = Version;
        header[5] = (byte)frameType;
        header[6] = 0;
        header[7] = 0;

        var transferIdBytes = Encoding.ASCII.GetBytes(transferId.ToString("N"));
        transferIdBytes.CopyTo(header[8..40]);
        BinaryPrimitives.WriteInt32LittleEndian(header[40..44], sequence);
        BinaryPrimitives.WriteInt64LittleEndian(header[44..52], offsetBytes);
        BinaryPrimitives.WriteInt32LittleEndian(header[52..56], payloadByteCount);
    }
}

File: src/ios/Sources/AnsightCore/PairingFileTransferWireProtocol.swift
import Foundation
import Network

public enum PairingFileTransferWireProtocol {
    public static let protocolName = "ansight.file-transfer.v1"
    public static let headerSize = 56
    public static let version: UInt8 = 1
    public static let magic = "ASFT"

    public static func createFrame(
        transferId: UUID,
        frameType: PairingFileTransferFrameType,
        sequence: Int32,
        offsetBytes: Int64,
        payload: Data
    ) -> Data {
        var frame = Data(count: headerSize + payload.count)
        writeHeader(
            into: &frame,
            transferId: transferId,
            frameType: frameType,
            sequence: sequence,
            offsetBytes: offsetBytes,
            payloadByteCount: Int32(payload.count)
        )
        frame.replaceSubrange(headerSize..<headerSize + payload.count, with: payload)
        return frame
    }

    public static func writeHeader(
        into header: inout Data,
        transferId: UUID,
        frameType: PairingFileTransferFrameType,
        sequence: Int32,
        offsetBytes: Int64,
        payloadByteCount: Int32
    ) {
        precondition(header.count >= headerSize, "The file transfer header buffer was too small.")
        let magicBytes = Array(magic.utf8)
        header[0] = magicBytes[0]
        header[1] = magicBytes[1]
        header[2] = magicBytes[2]
        header[3] = magicBytes[3]
        header[4] = version
        header[5] = frameType.rawValue
        header[6] = 0
        header[7] = 0

        let transferIdBytes = Array(transferId.uuidString.replacingOccurrences(of: "-", with: "").lowercased().utf8)
        header.replaceSubrange(8..<40, with: transferIdBytes.prefix(32))
        header.writeLittleEndian(sequence, at: 40)
        header.writeLittleEndian(offsetBytes, at: 44)
        header.writeLittleEndian(payloadByteCount, at: 52)
    }
}
